dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/helper/PrometheusBuilder30.java

```diff
@@ -30,47 +37,269 @@
+
+        /*
+         * If there exists global tags, then we will need to copy
+         *
+         */
+        if (globalTags != null && globalTags.length != 0) {
+            org.eclipse.microprofile.metrics.Tag[] applicationTags = mid.getTagsAsArray();
+            org.eclipse.microprofile.metrics.Tag[] combinedTags = new org.eclipse.microprofile.metrics.Tag[globalTags.length + applicationTags.length];
+
+            /*
+             * Copying Global Tags first ensures that any application tag
+             * with matching key will override the global tag when
+             * it is copied into the array.
+             *
+             * This is because the MetricID will read through the array
+             * and take the "newest" value.
+             *
+             * This is the behaviour in mpMetrics-2.x and below
+             */
+            System.arraycopy(globalTags, 0, combinedTags, 0, globalTags.length);
+            System.arraycopy(applicationTags, 0, combinedTags, globalTags.length, applicationTags.length);
+
+            MetricID temp = new MetricID("temporary", combinedTags);
+
+            return temp.getTagsAsString();
+        } else {
+            return mid.getTagsAsString();
+        }
+    }
+
+    public static void buildTimer30(StringBuilder builder, String name, String description, Map<MetricID, Metric> currentMetricMap) {
+        buildMetered30(builder, name, description, currentMetricMap);
         double conversionFactor = Constants.NANOSECONDCONVERSION;
 
         String lineName = name + "_elapsedTime_" + MetricUnits.SECONDS.toString();
         getPromTypeLine(builder, lineName, "gauge");
         for (MetricID mid : currentMetricMap.keySet()) {
-            getPromValueLine(builder, lineName, ((Timer) currentMetricMap.get(mid)).getElapsedTime().toNanos() * conversionFactor, mid.getTagsAsString());
+
+            getPromValueLine(builder, lineName, ((Timer) currentMetricMap.get(mid)).getElapsedTime().toNanos() * conversionFactor, resolveTagsAsStringWithGlobalTags(mid));
         }
 
         // Build Histogram
-        buildSampling(builder, name, description, currentMetricMap, conversionFactor, Constants.APPENDEDSECONDS);
+        buildSampling30(builder, name, description, currentMetricMap, conversionFactor, Constants.APPENDEDSECONDS);
     }
 
-    public static void buildSimpleTimer(StringBuilder builder, String name, String description, Map<MetricID, Metric> currentMetricMap) {
+    public static void buildSimpleTimer30(StringBuilder builder, String name, String description, Map<MetricID, Metric> currentMetricMap) {
         double conversionFactor = Constants.NANOSECONDCONVERSION;
 
-        buildCounting(builder, name, description, currentMetricMap);
+        buildCounting30(builder, name, description, currentMetricMap);
 
         String lineName = name + "_elapsedTime_" + MetricUnits.SECONDS.toString();
         getPromTypeLine(builder, lineName, "gauge");
```

Comment:
Would be nice to convert these string literals "gauge", "counter" to constants.

Reply:
Fixed.